### PR TITLE
Improved /vote further

### PR DESCRIPTION
**Pull request by MartinKrausewitz**
/vote has now the option to select the vote group in the command (if not declared it still uses the default vote from the config). The User has to be above (or equals) to the Group he wants to start a vote for (in the Groupranking). Alos the Buttons will be removed after the vote ends.
Also added /listgroups so the user can find which group they could start a vote for, also so the users can find the position of the group they want to start a vote for.

### CODE REVIEW

File: slashCommands/listgroups.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+export default {
+    command: {
+        name: "listgroups",
+        description: "This Command lists the groups of the server",
+        options: []
+    },
+    handler: async function(interaction: CommandInteraction) {
+        const roles = interaction.member.roles as GuildMemberRoleManager;
+        console.log(roles.cache)
```
Comment (c0derMo) [inline slashCommands/listgroups.ts:15]: Floating console.log

File: slashCommands/vote.ts [hunk 2]
```diff
@@ -29,6 +29,13 @@ export default {
                 description: "Time the vote is running in seconds.",
                 required: false,
                 minValue: 0
+            },
+            {
+                type: ApplicationCommandOptionType.Integer,
```
Comment (c0derMo) [inline slashCommands/vote.ts:34]: There is "ApplicationCommandOptionType.Role" (at least according to the documentation) - Wouldn't that be fitting here, then we could avoid all the role-integer stuff

File: slashCommands/vote.ts [hunk 3]
```diff
@@ -145,14 +163,26 @@ async function printVotes(pro: Set<string>, con: Set<string>, reply: Message, ti
         msgEmbed.setDescription(title);
     }
     //show embed
-    await reply.edit({embeds:[msgEmbed]})
+    await reply.edit({embeds:[msgEmbed], components:[]})
+}
+
+function playercanstartvote (interaction: CommandInteraction, messagerolepos: number): boolean {
```
Comment (c0derMo) [inline slashCommands/vote.ts:169]: - I think we should use the term "user"/"member" over "player"
- I think this method should either get a user, or just the list of roles, since it doesn't need anythine else

File: slashCommands/vote.ts [hunk 4]
```diff
@@ -145,14 +163,26 @@ async function printVotes(pro: Set<string>, con: Set<string>, reply: Message, ti
...
+    roles.cache.forEach((r)=> {
+        if (r.position > phrp) {
+            phrp = r.position;
+        }
+    });
+    return messagerolepos > phrp
 }
 
-function getEmbedOptions(title:string, msg: string, timestamp?:number): BaseMessageOptions {
+function getEmbedOptions(title:string, msg: string, group:number, timestamp?:number): BaseMessageOptions {
```
Comment (c0derMo) [inline slashCommands/vote.ts:180]: As far as I'm aware, because of the large size of id's in discord, it might be too large for a regular integer. While I dont know how js handles that, I think we should just use discord.js's `Snowflake` (which is just their fancy word for their IDs) for things like this

Comment (MartinKrausewitz) [reply slashCommands/vote.ts]: javascript only uses 53 bit. python wouldnt have this problem ^^

**Comment by MartinKrausewitz** [comment]
Everything fixed. Please review again @c0derMo @myphyk @Kr0nox

**Comment by MartinKrausewitz** [comment]
/listgroups may now be redundant, because you can now see in the vote command the options for the role, but you see all roles not just the ones you could start a vote for.

File: slashCommands/vote.ts [hunk 5]
```diff
@@ -31,7 +39,7 @@ export default {
                 minValue: 0
             },
             {
-                type: ApplicationCommandOptionType.Integer,
+                type: ApplicationCommandOptionType.Role,
                 name: "votegroup",
                 description: "The Group that is allowed to vote (0 for @everyone)",
```
Comment (c0derMo) [inline slashCommands/vote.ts:44]: `0 for @everyone` is still remaining from when this was an integer - since `@everyone` is just selectable now, I think this notice can go

**Review by c0derMo — COMMENTED** [review]
I'd also guess that the `/listgroups` command isn't necessary anymore
While it does reveal all roles below your highest, since we're checking for role equality, and not all roles below, starting a vote for a role that you don't own seems off, because it wouldn't allow yourself to vote, and all your current roles are always visible to you through discord directly.

Tl;dr:
I think we can get rid of /listgroups, and maybe even enforce that the user needs to have the role he wants to start a vote for

**Comment by MartinKrausewitz** [comment]
The user now needs to have the group he wants to start a vote for. Listgroups is removed and the Messages are fixed. I think we are ready to merge @c0derMo @Kr0nox @myphyk

File: slashCommands/vote.ts [hunk 6]
```diff
@@ -43,9 +58,16 @@ export default {
...
         const title = options.getString("name", true);
+        let maingroup = options.getRole("votegroup", false) as Role;
+        if (maingroup == undefined) {
+            const guid = get("vote_group", "config") as string;
+            maingroup = (guid == "")? interaction.guild.roles.everyone: interaction.guild.roles.cache.get(guid);
+        }
 
-        const guid = get("vote_group", "config") as string;
-        const maingroup = interaction.guild.roles.cache.get(guid);
+        if (membercanstartvote(interaction.member as GuildMember, maingroup)) {
```
Comment (c0derMo) [inline slashCommands/vote.ts:67]: This fails when the role is `@everyone` since that role doesn't have an id and `membercanstartvote` tries to access it... I think we should check for everyone role here (by comparison) and only execute the function if the role isnt everyone
(e.g. `!maingroup.equals(everyone) && membercanstartvote`)

File: slashCommands/vote.ts [hunk 7]
```diff
@@ -145,14 +167,20 @@ async function printVotes(pro: Set<string>, con: Set<string>, reply: Message, ti
         msgEmbed.setDescription(title);
     }
     //show embed
-    await reply.edit({embeds:[msgEmbed]})
+    await reply.edit({embeds:[msgEmbed], components:[]})
+}
+
+function membercanstartvote (member: GuildMember, selectedRole: Role): boolean {
```
Comment (c0derMo) [inline slashCommands/vote.ts:173]: The name of this function is a bit misleading at this point, because it checks if a member can't start a vote, while it would indicate the negation of this by its name

File: slashCommands/vote.ts (hunk 6 above)
Comment (MartinKrausewitz) [reply slashCommands/vote.ts:67]: the snowflake should be created at server creation? The id of a role isnt a optional so @everyone has to have one

Comment (MartinKrausewitz) [reply slashCommands/vote.ts]: and it works just fine on my instance

Comment (c0derMo) [reply slashCommands/vote.ts]: You're right, specifying everyone as the vote-group works, it didn't work for me when the `vote_group` isn't set at all in the config. This is actually partially a problem with the underlying configmanager, because instead of returning an empty string or `undefined` when trying to look up an object that doesn't exist, we currently return `{}`.
As a temporary workaround until I fix the config manager, we can change the condition in Line 64 to `JSON.stringify(guid) == "{}"`.

**Comment by MartinKrausewitz** [comment]
Now everything should be fine ;D @c0derMo @Kr0nox @myphyk

**Review by c0derMo — APPROVED** [review]
lgtm
